server/src/main/java/io/crate/auth/HttpAuthUpstreamHandler.java

```diff
@@ -95,7 +99,26 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Except
     private void handleHttpRequest(ChannelHandlerContext ctx, HttpRequest request) {
         SSLSession session = getSession(ctx.channel());
         Credentials credentials = credentialsFromRequest(request, session, settings);
-        String username = credentials.username();
+
+        String username;
+        var decodedJwt = credentials.decodedToken();
+        if (decodedJwt != null) {
+            Role role = roles.findUser(
+                decodedJwt.getIssuer(),
+                decodedJwt.getClaim("username").asString()
+            );
+            if (role != null) {
+                username = role.name();
+            } else {
+                // Set to empty string to avoid NPE
+                // and also align with behavior of Header.extractCredentialsFromHttpAuthHeader where we set an empty user as fallback.
+                username = "";
+            }
+            credentials.setUsername(username);
+        } else {
+            username = credentials.username();
+        }
+        assert username != null : "Username must be either resolved or set to empty";
         if (username.equals(authorizedUser)) {
```

Comment:
Good catch!

I see 2 ways to address that:
1. Don’t store authorizedUser for JWT. It would also mean that we try authentication on each access which might be redundant for the time frame where token is valid.
 2.   Keep storing authorizedUser but also store expiration date per session  - and we can renew session only when needed. Bit more complex but more optimal. 
 
 I think I should do 1 now and follow up on 2 to keep the scope of the current PR smaller. wdyt?

Reply:
As discused internally, I went with per-request check
https://github.com/crate/crate/pull/15551/commits/f446957c59826024aa794257c9e1302d247caf72

Haven't found anything specific in RFC, but in general JWT is [stateless authentication](https://en.wikipedia.org/wiki/JSON_Web_Token) so I went with not storing any state/authorizedUser